src/components/Swap/SwapButton/useSwitchNetwork.tsx

```diff
@@ -0,0 +1,41 @@
+import { Trans } from '@lingui/macro'
+import { Action } from 'components/ActionButton'
+import useActiveWeb3React from 'hooks/useActiveWeb3React'
+import { Spinner } from 'icons'
+import { useCallback, useMemo, useState } from 'react'
+
+export default function useSwitchNetwork(desiredChainId?: number) {
+  const { chainId, connector } = useActiveWeb3React()
+
+  const [isPending, setIsPending] = useState(false)
+  const onSwitchNetwork = useCallback(async () => {
+    setIsPending(true)
+    try {
+      const desiredChainIdHex = `0x${desiredChainId?.toString(16)}`
+      await connector.provider
+        ?.request({
+          method: 'wallet_switchEthereumChain',
+          params: [{ chainId: desiredChainIdHex }],
+        })
+        .then(() => connector.activate(desiredChainId))
```

Comment:
1. How does this handle if a user does not have the chain info added to their wallet yet? 
2. Have you taken a look at the code to accomplish this on the interface? I'm familiar w/ the approach you've taken here, but it looks like there's a new way to do this w/ the current version of web3-react? 

```
export const switchChain = async (connector: Connector, chainId: SupportedChainId) => {
  if (!isChainAllowed(connector, chainId)) {
    throw new Error(`Chain ${chainId} not supported for connector (${typeof connector})`)
  } else if (connector === walletConnectConnection.connector || connector === networkConnection.connector) {
    await connector.activate(chainId)
  } else {
    const info = getChainInfo(chainId)
    const addChainParameter = {
      chainId,
      chainName: info.label,
      rpcUrls: getRpcUrls(chainId),
      nativeCurrency: info.nativeCurrency,
      blockExplorerUrls: [info.explorer],
    }
    await connector.activate(addChainParameter)
  }
}
```

@vm thoughts?

Reply:
you can just call `connector.activate(chainId)` now instead of doing the provider request

Reply:
and you'll need to pass the chainInfo! very similar to interface code.

Reply:
small problem I only upgraded the w3r version in the `wallet-connect-flow` branch (which hasn't been merged yet) - and it comes with a couple changes that break how we use web3react currently

Reply:
current version of w3r doesn't support activate(chainId)

Reply:
can't we merge that first and then rebase this?